Use dropdown without auto completion for simple lists in edit hosting preferences form

thdk opened the pull request:
<!---
Please describe the pull request below.
If it closes an issue, make sure to write "closes #1234"
If there is an issue but it isn't completely closed, still refer to the issue number, eg. "part of #1234"
--->

Closes #1268 

Sleeping arrangement (or 'private space'), smoking location and parking details fields on the edit hosting preferences form now no longer allow typing text with autocompletion.

These fields now display the valid options in a dropdown (MUI select component)

Users may have 0 (unspecified) or 1 (unknown) as value for these fields. 
Before this issue, this would log warnings in the console because they were not listed as an option for the autocomplete control.

I have now added the unknown (ask me) option to the dropdown for each of these fields.
If the value of the user was 0 (unspecified) it will fallback in the UI to 1 (unknown) and be saved as 1 (unknown) if the user would submit the form.

The user may now also change a previous saved value to 'ask me'.

If this is not the expected behavior, please let me know.

<!---
Checklists - you can remove one that is not applicable (ie. remove backend checklist if you only worked on frontend)
If you need help with any of these, please ask :)
--->

**Frontend checklist**
- [x] Formatted my code with `yarn format && yarn lint --fix`
- [x] There are no warnings from `yarn lint`
- [x] There are no console warnings when running the app
- [x] Added any new components to storybook
- [ ] Added tests where relevant
- [x] All tests pass
- [x] Clicked around my changes running locally and it works
- [x] Checked Desktop, Mobile and Tablet screen sizes


<!---
Remember to request review from couchers-org/frontend, couchers-org/@backend or an individual.
Once your code is approved, remember to merge it if you have write access
--->

thdk committed:
feat: add select component

thdk committed:
feat: use Select component on edit hosting preferences page

fixes #1268

thdk committed:
fix: use the correct label for sleeping arrangement

darrenvong commented on app/frontend/src/components/Select.tsx:
```diff
@@ -0,0 +1,53 @@
+    >
+      <InputLabel htmlFor={id}>{label}</InputLabel>
+      <MuiSelect
+        native
```
I wonder if it's better to not use the `native` option so we get a consistent look and feel between operating systems and also against our other components

thdk replied:
"As the user experience can be improved on mobile using the native select of the platform, we allow such pattern."
https://material-ui.com/components/selects/#native-select

darrenvong reviewed (commented):
LGTM! Maybe the native point I raised should be relayed to Slack to ask for some input from the design team...

lucaslcode commented on app/frontend/src/features/profile/edit/EditHostingPreference.tsx:
```diff
@@ -53,7 +54,7 @@ import {
   SmokingLocation,
 } from "pb/api_pb";
-import { useState } from "react";
+import React, { useState } from "react";
```
revert

lucaslcode commented on app/frontend/src/features/profile/edit/EditHostingPreference.tsx:
```diff
@@ -95,6 +96,7 @@ const useStyles = makeStyles((theme) => ({
     paddingTop: theme.spacing(1),
   },
   field: {
+    display: "block",
```
Is this necessary? It is also in your new `<Select>` component

lucaslcode commented on app/frontend/src/features/profile/edit/EditHostingPreference.tsx:
```diff
@@ -250,31 +252,25 @@ export default function HostingPreferenceForm() {
+                defaultValue={user.smokingAllowed || SmokingLocation.SMOKING_LOCATION_UNKNOWN}
+                onChange={onChange}
+                label={ACCEPT_SMOKING}
+                name="smokingAllowed"
```
```suggestion
```

Name is set in the controller

lucaslcode commented on app/frontend/src/features/profile/edit/EditHostingPreference.tsx:
```diff
@@ -250,31 +252,25 @@ export default function HostingPreferenceForm() {
-                  />
-                )}
-              />
+                ].map((o) => (
+                  <option value={o} key={o}>
+                    {smokingLocationLabels[o]}
+                  </option>
```
Instead of doing this every time, what do you think of giving `<Select>` an `options: {label: string, value: T}[]` prop and doing the map in the component?

thdk replied:
will do :)

lucaslcode commented on app/frontend/src/features/profile/edit/EditHostingPreference.tsx:
```diff
@@ -250,31 +252,25 @@ export default function HostingPreferenceForm() {
-                multiple={false}
-                options={[
+              <Select
+                defaultValue={user.smokingAllowed || SmokingLocation.SMOKING_LOCATION_UNKNOWN}
```
Seems like code smell to have `defaultValue` on both the controller and the rendered component. Maybe the most correct way would be to set `value` on the component from `render={({ onChange, value }) => (` ? Not totally sure

thdk replied:
yes :( seems it didn't work indeed. I've added tests now to check if the correct value from the database is set in the form.

thdk committed:
style: remove id prop

the id prop isn't used in this component and will be passed to the child component anyway as part of otherProps (if not destructured)

thdk committed:
add 'for' attribute to select label

This is needed so we can target select component by its label text in tests

thdk committed:
test: add tests for field values of hosting preferences form

thdk committed:
fix default value issue + review feedback

thdk committed:
style: formatting

thdk committed:
feat: add options prop to select component

lucaslcode commented on app/frontend/src/features/profile/edit/EditHostingPreference.tsx:
```diff
@@ -249,28 +248,55 @@ export default function HostingPreferenceForm() {
+                      smokingLocationLabels[
+                        SmokingLocation.SMOKING_LOCATION_NO
+                      ],
+                  },
```
Haha this is worse...
What if you do `Object.entries(smokingLocationLabels).map(([key, value]) => ({label: value, value: +key}))`?
However that may be  less type safe and one could argue it's more opaque.

You decide if you want to leave it like you have done now, how it was before, or that one-liner.

thdk replied:
I agree it looks worse, but it is more 'solid' because at least you don't have to write html tags (but still can using the children prop of select). I've tried it with object.entries but then we are less flexibile in the sort order of options and you can't have entries in the labels which you don't want to display in the dropdown. 
Eg. the labels contains both a label for unknown and unspecified while now I only show unknown in the dropdown.

Another option is to pass the select component an array of values and a dictionary (or a getLabel function) from which to retrieve the label for each value.

lucaslcode replied:
I think that's a good idea - our usual use case is the protobuf enums, and in other cases, it probably won't be much more verbose than any other way

lucaslcode reviewed (changes requested):
One comment - if you want to leave as is, re-request and I'll accept

thdk committed:
feat: add data property to select component to provide labels for each option

lucaslcode commented on app/frontend/src/components/Select.tsx:
```diff
@@ -20,19 +20,41 @@ const useStyles = makeStyles((theme) => ({
   const classes = useStyles();
+
+  const SelectOptions = () => {
+    if (!options || !data) {
```
If this is the case, those props shouldn't be optional and you won't need this

thdk replied:
You can still provide options using the `children `prop. In that case `options `and `data `can be undefined.

lucaslcode replied:
Is that necessary?

thdk replied:
no, but it's cool :) (I'll remove it)

lucaslcode commented on app/frontend/src/components/Select.tsx:
```diff
@@ -20,19 +20,41 @@ const useStyles = makeStyles((theme) => ({
-  options?: { label: string; value: string | number }[];
+  options?: Extract<keyof T, string | number>[];
+  value?: T extends undefined ? string | number : keyof T;
+  data?: T;
```
If you make the type `T extends Record<string | number, string>`, maybe you don't need `options` at all?

And what do you think about making the name `labelledOptions` or `optionsAndLabels` or something?

thdk replied:
The reason I need both `options` and `data` is because the labels object contains more items than I want to display in the dropdown. I don't see how that would be different when using Record (which seems a better option here indeed) than an index signature.

lucaslcode replied:
Oh, I see. Can you think of a better name for `data`? I wouldn't know the difference between it and `options` without looking at a previous usage or the source.

thdk replied:
what about 'optionLabelMap' ?

lucaslcode replied:
Sounds good

darrenvong commented on app/frontend/src/components/TextField.tsx:
```diff
@@ -37,7 +36,6 @@ export default function TextField({
     <MuiTextField
       {...otherProps}
       variant={variant}
-      id={id}
```
Why was this removed? We need this previously so [MUI correctly links label up with the right ID](https://material-ui.com/components/text-fields/#accessibility). Removing it will break this

thdk replied:
I've also removed  `id` from the destructured props. So `otherProps` will still have the `id` prop. The spread operator used on line 38 will nicely pass all these props, including the `id` prop, to the MuiTextField component.

So this change is just a 'code style' issue.

darrenvong commented on app/frontend/src/components/Select.tsx:
```diff
@@ -0,0 +1,79 @@
+}) {
+  const classes = useStyles();
+
+  const SelectOptions = () => {
```
Can we lift this out of this component? Creating a component inside a component is an anti-pattern as it creates a new function reference each time so React won't know it's the same component and so screws up its reconciliation and ends up recreating the component tree and replacing more things in the DOM than necessary

thdk replied:
I can lift this out of the component. But I can also use useMemo for this. Any preference?

lucaslcode replied:
Moving it out is better, but in this case you can just inline the options instead of having another component

thdk committed:
review feedback

thdk committed:
feedback: inline options

lucaslcode reviewed (approved):
Great! Pretty sure all Darren's comments are addressed too so I'll approve :)
